File: 11-04/6/ac/ethan.cpp
#include <iostream>
#include <cstring>

#define MAX_N 505
#define MAX_B 102

using namespace std;

int T, N, B[MAX_N];
bool dp[MAX_N][MAX_N*MAX_B];

int main() {
    cin >> T;
    while (T--) {
        memset(dp, 0, sizeof(dp));
        cin >> N;
        int total = 0;
        for (int i = 1; i <= N; i++) {
            cin >> B[i];
            total += B[i];
        }
        dp[0][0] = true;
        for (int i = 1; i <= N; i++) {
            for (int j = 0; j <= i*MAX_B; j++) {
                dp[i][j] = dp[i - 1][j];
                if (j >= B[i]) {
                    dp[i][j] |= dp[i - 1][j - B[i]];
                }
            }
        }
        int best = total;
        for (int j = 0; j <= N * MAX_B; j++) {
            if (dp[N][j])
                best = min(best, abs(total - 2*j));
        }
        cout << best << '\n';
    }
    
    return 0;
}

File: 11-04/6/ac/ethan.py
MAX_B = 75

T = int(raw_input())

for t in range(T):
    N = int(raw_input())
    B = map(int, raw_input().split())
    # dp[i][j] = True iff there is a subset of the first i pins with sum j
    dp = [[False]*(N*MAX_B + 1) for _ in range(N + 1)]
    dp[0][0] = True
    for i in range(1, N + 1):
        for j in range(i*MAX_B + 1):
            dp[i][j] = dp[i - 1][j]
            if j >= B[i - 1] and dp[i - 1][j - B[i - 1]]:
                dp[i][j] = True
    total = sum(B)
    best = sum(B)
    for j in range(N*MAX_B + 1):
        if dp[N][j]:
            best = min(best, abs(total - 2 * j))
    print best
